escape_roomba/thread_manager.py

```diff
@@ -0,0 +1,327 @@
+
+    #
+    # Internal methods
+    #
+
+    async def _async_maybe_update(self, channel_id=None, channel=None,
+                                  message=None, message_id=None, emoji=None):
+        """Responds to an indication of message change; if the change is
+        relevant, fetches and processes the message's latest version.
+        One each of channel_id/channel and message_id/message must be non-None.
+
+        Args:
+            channel_id: int - ID of channel containing message *OR*
+            channel: discord.TextChannel - channel object containing message
+            message_id: int - ID of message to fetch and process *OR*
+            message: discord.Message - contents of updated message
+            emoji: str, discord.*Emoji - emoji of reaction change (or None)
+        """
+
+        # Handle various ways channel and message can be supplied
+        ci = channel_id or channel.id
+        mi = message_id or message.id
+
+        # Skip irrelevant changes; a change is "relevant" if:
+        # - there is an existing fetch in progress for the message OR
+        # - the message is an existing thread channel origin OR
+        # - the message has a thread emoji reaction OR
+        # - there was a change to thread emoji reactions
+        if ((ci, mi) not in self._update_done and
+            mi not in self._thread_by_origin.get(ci, {}) and
+            (emoji is None or str(emoji) != self._THREAD_EMOJI) and
+            (message is None or
+             all(str(r) != self._THREAD_EMOJI for r in message.reactions))):
+            return
+
+        # To avoid race conditions with asynchronous message fetch requests,
+        # serialize update operations, using a per-message-ID map of Futures.
+        # (Per-message Locks would be easier, but would accumulate over time.)
+        last_update_done = self._update_done.get((ci, mi))
+        this_update_done = self._loop.create_future()
+        self._update_done[(ci, mi)] = this_update_done
+        if last_update_done is not None:
+            await last_update_done
+
+        channel = channel or self._context.client.get_channel(channel_id)
+        if channel:
+            self._logger.debug(f'Fetch m=%s ...', fid(mi))
+            try:
+                fetched_message = await channel.fetch_message(mi)
+                await self._async_process_message(fetched_message)
+            except discord.errors.NotFound:
+                self._logger.debug('Fetch m=%s: Gone', fid(mi))
+                await self._async_message_gone(channel_id=ci, message_id=mi)
+        else:
+            # No channel available, treat the message as missing.
+            self._logger.debug('Fetch m=%s: !Chan', fid(mi))
+            await self._async_message_gone(channel_id=ci, message_id=mi)
+
+        # If another fetch is waiting to run, let it go.
+        if self._update_done.get((ci, mi)) is not this_update_done:
+            this_update_done.set_result(True)  # Let the next fetch run.
```

Comment:
Dunno if this would be better or worse, but I wonder if it might be more understandable to always do a  `this_update_done.set_result(True)` at the end of this method, and at the beginning of the method (or maybe in the same block that waits for `last_update_done`?) remove the `done` future from the map. Not sure if it's worth the trouble.

Reply:
Done.